[UnifiedPDF] [macOS] Switching from continuous to discrete mode will display the page(s) that are at the top of the window, even if barely visible

aprotyas opened the pull request:
#### 7242348f028d45f8a1415f3cfe0b7d40c90de1da
<pre>
[UnifiedPDF] [macOS] Switching from continuous to discrete mode will display the page(s) that are at the top of the window, even if barely visible
<a href="https://bugs.webkit.org/show_bug.cgi?id=286193">https://bugs.webkit.org/show_bug.cgi?id=286193</a>
<a href="https://rdar.apple.com/137608841">rdar://137608841</a>

Reviewed by Tim Horton.

When switching between display modes, we need to decide an anchor point
to maintain the PDF document&apos;s view position post factum. We have been
using the top left (plugin space) as said anchor point, but when
switching out from a continuous display mode, the anchor point often
happens to be in the sliver of the previous page showing in the plugin
space, so after the transition, we end up landing on a previous page.

This patch addresses this logic error by instead adopting Preview&apos;s (and
PDFLayerController&apos;s) behavior of anchoring around the origin (top left)
of the page visible in the center of the plugin.

To facilitate these changes, we introduce a couple of helper methods,
namely `anchorPointForDocument()` and `pageIndexForCurrentView()`. These
helpers allow both the presentation controller instances to use shared
logic for `pdfPositionForCurrentView()`, so we devirtualize it. Note
that we don&apos;t use this new anchoring behavior for layout updates with
autoscaling behavior activated, since we still want to unconditionally
anchor around the plugin&apos;s top-left in that use case.

* LayoutTests/pdf/two-pages-continuous-to-discrete-expected.html:

Address some flakiness.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
(WebKit::PDFDiscretePresentationController::pageIndexForCurrentView const):
(WebKit::PDFDiscretePresentationController::pdfPositionForCurrentView const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.mm:
(WebKit::PDFPresentationController::pdfPositionForCurrentView const):
(WebKit::PDFPresentationController::anchorPointForDocument const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.mm:
(WebKit::PDFScrollingPresentationController::pageIndexForCurrentView const):
(WebKit::PDFScrollingPresentationController::pdfPositionForCurrentView const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::updateLayout):
(WebKit::UnifiedPDFPlugin::setDisplayModeAndUpdateLayout):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UnifiedPDFWithKeyboardScrolling::synchronouslyLoadPDFDocument):
(TestWebKitAPI::UnifiedPDFWithKeyboardScrolling::scrollDown):
(TestWebKitAPI::UnifiedPDFWithKeyboardScrolling::scrollRight):
(TestWebKitAPI::UnifiedPDFWithKeyboardScrolling::pressKey):
(TestWebKitAPI::TEST_F(UnifiedPDFWithKeyboardScrolling, KeyboardScrollingInSinglePageMode)):
(TestWebKitAPI::TEST_F(UnifiedPDFWithKeyboardScrolling, DisplayModeTransitionLandingPage)):

Introduce a UnifiedPDFWithKeyboardScrolling text fixture class, which
takes the keyboard scrolling technique from the existing API test -
KeyboardScrollingInSinglePageMode - and exposes it as convenient
`scroll[Down|Right]()` utilities. This fixture class is used to
introduce a new API test `DisplayModeTransitionLandingPage`.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/multiple-pages-colored.pdf: Added.

Canonical link: <a href="https://commits.webkit.org/289212@main">https://commits.webkit.org/289212@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ecb51a10e1af98e6cca536acc482b00916cd097f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85726 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40132 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90806 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36714 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87767 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5626 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13395 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66612 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24419 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88731 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4329 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77823 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46897 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4189 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32093 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35792 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74835 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32947 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/92563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13045 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9577 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75356 "35 flakes 185 failures") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13255 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73657 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74498 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18376 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18713 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17149 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/5153 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13071 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18433 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12860 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16286 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14647 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
```diff
@@ -85,49 +85,97 @@ - (bool)navigationFinished
+    static constexpr unsigned short DownArrowKeyCode { 0x7D };
+    static constexpr unsigned short RightArrowKeyCode { 0x7C };
+
+    void pressKey(auto key, unsigned short code, Seconds duration = 200_ms)
```
where is pressKey used? I'm having trouble finding it

aprotyas replied:
The Github diff viewer doesn't help here at all, but it's used in `scrollDown()` and `scrollRight()`.

rr-codes replied:
oh nvm found it

hortont424 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
```diff
@@ -85,49 +85,97 @@ - (bool)navigationFinished
+
+    auto colorsBefore = [webView sampleColors];
+
+    scrollDown(800_ms);
```
800ms!!

aprotyas replied:
Indeed, I guess we can consider setting the scroll position programmatically.

aprotyas committed:
[UnifiedPDF] [macOS] Switching from continuous to discrete mode will display the page(s) that are at the top of the window, even if barely visible
https://bugs.webkit.org/show_bug.cgi?id=286193
rdar://137608841

Reviewed by Tim Horton.

When switching between display modes, we need to decide an anchor point
to maintain the PDF document's view position post factum. We have been
using the top left (plugin space) as said anchor point, but when
switching out from a continuous display mode, the anchor point often
happens to be in the sliver of the previous page showing in the plugin
space, so after the transition, we end up landing on a previous page.

This patch addresses this logic error by instead adopting Preview's (and
PDFLayerController's) behavior of anchoring around the origin (top left)
of the page visible in the center of the plugin.

To facilitate these changes, we introduce a couple of helper methods,
namely `anchorPointForDocument()` and `pageIndexForCurrentView()`. These
helpers allow both the presentation controller instances to use shared
logic for `pdfPositionForCurrentView()`, so we devirtualize it. Note
that we don't use this new anchoring behavior for layout updates with
autoscaling behavior activated, since we still want to unconditionally
anchor around the plugin's top-left in that use case.

* LayoutTests/pdf/two-pages-continuous-to-discrete-expected.html:

Address some flakiness.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDiscretePresentationController.mm:
(WebKit::PDFDiscretePresentationController::pageIndexForCurrentView const):
(WebKit::PDFDiscretePresentationController::pdfPositionForCurrentView const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.mm:
(WebKit::PDFPresentationController::pdfPositionForCurrentView const):
(WebKit::PDFPresentationController::anchorPointForDocument const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.mm:
(WebKit::PDFScrollingPresentationController::pageIndexForCurrentView const):
(WebKit::PDFScrollingPresentationController::pdfPositionForCurrentView const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::updateLayout):
(WebKit::UnifiedPDFPlugin::setDisplayModeAndUpdateLayout):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UnifiedPDFWithKeyboardScrolling::synchronouslyLoadPDFDocument):
(TestWebKitAPI::UnifiedPDFWithKeyboardScrolling::scrollDown):
(TestWebKitAPI::UnifiedPDFWithKeyboardScrolling::scrollRight):
(TestWebKitAPI::UnifiedPDFWithKeyboardScrolling::pressKey):
(TestWebKitAPI::TEST_F(UnifiedPDFWithKeyboardScrolling, KeyboardScrollingInSinglePageMode)):
(TestWebKitAPI::TEST_F(UnifiedPDFWithKeyboardScrolling, DisplayModeTransitionLandingPage)):

Introduce a UnifiedPDFWithKeyboardScrolling text fixture class, which
takes the keyboard scrolling technique from the existing API test -
KeyboardScrollingInSinglePageMode - and exposes it as convenient
`scroll[Down|Right]()` utilities. This fixture class is used to
introduce a new API test `DisplayModeTransitionLandingPage`.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/multiple-pages-colored.pdf: Added.

Canonical link: https://commits.webkit.org/289212@main
